service/api_task.go

```diff
@@ -566,15 +537,28 @@ func assignNextAvailableTask(ctx context.Context, taskQueue *model.TaskQueue, di
 			"task_project":            nextTask.Project,
 			"task_group_max_hosts":    nextTask.TaskGroupMaxHosts,
 		})
-
 		if ok && isTaskGroupNewToHost(currentHost, nextTask) {
-			numHosts, err := host.NumHostsByTaskSpec(nextTask.TaskGroup, nextTask.BuildVariant, nextTask.Project, nextTask.Version)
-			if err != nil {
-				return nil, false, errors.WithStack(err)
+			dispatchRace := ""
+
+			// in this case, the task group is new to the host, but this isn't the first task in the
+			// single host task group, so we have a race.
+			if nextTask.TaskGroupMaxHosts == 1 && nextTask.TaskGroupOrder > 1 {
```

Comment:
What about the case where someone restarts a task in a task group that isn't the first one? I suppose they shouldn't, but they might. Could that cause the task to be stuck in the queue?

Reply:
😮  that is an interesting point. Yes? I think so. So either we could:
1) Block this work on https://jira.mongodb.org/browse/EVG-5313 (and maybe re-prioritize this now that it seems MCI-3200 is done?)
2) Check when restarting a task that it's taskGroupOrder isn't greater than 1 if single host task group (this should be okay performance-wise since we won't have to look up the other tasks in the task group)
3) Try to fix race without relying on order (I'm sort of unconvinced)

Reply:
I think #1 makes sense. I don't remember what the approach to #1 was, but from the comments it looks like we were parsing YAML. I wonder if we can bypass parserProject altogether, since we store task group info in the database now. I don't remember why we didn't try that approach in the first place. Maybe we hadn't started caching this metadata?

Reply:
The original PR needed parsing YAML. The post-parser-project PR is much better https://github.com/evergreen-ci/evergreen/pull/3401/files but we wanted to create an index on BuildId/TaskGroup which is why we were blocked.

Reply:
Can you estimate the size of the index?

Reply:
Hmm I'm looking at other indexes in the task collection... the other index involving Build ID is  28 GB but that's Build ID & ID which seems bigger, so I'd estimate 2 or 3 GB given what other indexes look like.

Thinking back distantly to M101, can't queries still use the first field in a compound index? Would the existing index on Build ID & ID help us filter down to all tasks in the build? That might be enough.

Reply:
I think tens of gigabytes is fine. What's the ideal index you would want to add?

I opened https://jira.mongodb.org/browse/BUILD-11381 to reclaim space.

Reply:
It would be an index on the task collection for Build ID and Task Group--this is the query that would be added in EVG-5313: 

https://github.com/ablack12/evergreen/blob/9a3570814c132116c3e78453beb7e24743fd3c2d/model/task/db.go#L868

Reply:
That seems fine to me.

Reply:
Is this something I need to wait on the build team for or can I just go ahead with re-opening that ticket and making the index?

Reply:
Not sure, I will look into it.

Reply:
The prod notes have 2.5-3TB in free space, and we can reclaim space as well, so I think building an index that makes tens of gigabytes is something we can do without waiting on build. Let's do it.

Reply:
@bsamek since EVG-5313 is merged we can merge this now if you have no other changes.

Reply:
I think the code looks good. I'm curious about one more case. Suppose task 1 of 5 finishes. After task 2 finishes and task 3 is running, a user restarts task 2. What will happen as a result of EVG-5313?

Reply:
Task 2 will be marked as reset in the backend, but won't restart until the whole task group is finished (so either task 3 fails and task 4/5 are blocked, task 3/4/5 succeeds, task 3, 4, or 5 are deactivated, etc).

I would like this to be more apparent  to users but I wasn't sure how on the old task page (/ if it was worth it). I can make a ticket to make it clear on the new task page?

Reply:
Yeah, that makes sense to me. LGTM.